app/forms/schedule_form.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+class ScheduleForm
+  include ActiveModel::Model
+
+  attr_reader :errors, :flash_success
+
+  def initialize(params, user, schedule = nil)
+    @errors = ActiveModel::Errors.new self
+
+    # Not strong
+    @params = params
+    @current_user = user
+    @schedule = schedule
+  end
+
+  def save
+    editable_order_cycle_ids_for_create
+
+    return false if @params[:order_cycle_ids].blank?
+
+    @schedule.attributes = permitted_resource_params
+
+    if @schedule.save
+      @schedule.order_cycle_ids = @params[:order_cycle_ids]
+      @schedule.save!
+
+      sync_subscriptions_for_create
+
+    end
+
+    true
+  end
+
+  def update(_params)
+    editable_order_cycle_ids_for_update
+
+    return false unless @schedule.update(permitted_resource_params)
+
+    sync_subscriptions_for_update
+  end
+
+  private
+
+  def editable_order_cycle_ids_for_create
+    return unless @params[:order_cycle_ids]
+
+    @existing_order_cycle_ids = []
+    result = editable_order_cycles(@params[:order_cycle_ids])
+    @params[:order_cycle_ids] = result
+  end
+
+  def editable_order_cycle_ids_for_update
+    return unless @params[:schedule][:order_cycle_ids]
+
+    @existing_order_cycle_ids = @schedule.order_cycle_ids
+    result = editable_order_cycles(@params[:schedule][:order_cycle_ids])
+
+    @params[:schedule][:order_cycle_ids] = result
+    @schedule.order_cycle_ids = result
+  end
+
+  def editable_order_cycles(requested)
+    permitted = OrderCycle
+      .where(id: @params[:order_cycle_ids] | @existing_order_cycle_ids)
+      .merge(OrderCycle.managed_by(@current_user))
+      .pluck(:id)
+    result = @existing_order_cycle_ids
+    result |= (requested & permitted) # add any requested & permitted ids
+    # remove any existing and permitted ids that were not specifically requested
+    result -= ((result & permitted) - requested)
+    result
+  end
+
+  def sync_subscriptions_for_create
+    return unless @params[:order_cycle_ids]
+
+    sync_subscriptions
+  end
+
+  def sync_subscriptions_for_update
+    return unless @params[:schedule][:order_cycle_ids] && @schedule.errors.blank?
+
+    sync_subscriptions
+  end
+
+  def sync_subscriptions
```

Comment:
Okay, it may need to go into a service or concern, though. ClassLength limit is currently 100 lines. I think we will be over. That's the main reason I packaged it in the form. It's also related to `existing_order_cycle_ids` which would need to be passed to it.

Reply:
The ClassLength limit is an arbitrary rule, it mostly there to prompt us to take another look if the file exceed the limit, it probably means the controller has become too complicated/is doing too much and maybe we need to refactor. 
But that shouldn't be a reason to move some code to somewhere it doesn't logically belong to. 
In this case, a service looks like a good solution. I had a quick look, it looks like you can derive `existing_order_cycle_ids` from `shedule`, so you should be able to do the same outside the `ScheduleForm`

Reply:
OK, thanks for the feedback and taking the time to explain! I moved the subscriptions back into the `schedules_controller.rb`.